File: sdk/python/src/roset/resources/org.py
from __future__ import annotations

from roset.http_client import HttpClient
from roset.models import ApiKey, Invitation, Member, Tenant


class OrgResource:
    """Organization/Tenant Management."""

    def __init__(self, http: HttpClient):
        self.http = http

    def get_tenant(self) -> Tenant:
        """Get current tenant details."""
        data = self.http.request("GET", "/v1/org")
        return Tenant.model_validate(data["tenant"])

    def list_members(self) -> list[Member]:
        """List workspace members."""
        data = self.http.request("GET", "/v1/org/members")
        return [Member.model_validate(item) for item in data.get("members", [])]

    def invite_member(self, email: str, role: str) -> Invitation:
        """Invite a new member."""
        data = self.http.request("POST", "/v1/org/invites", json={"email": email, "role": role})
        return Invitation.model_validate(data["invitation"])

    def list_api_keys(self) -> list[ApiKey]:
        """List API keys."""
        data = self.http.request("GET", "/v1/org/api-keys")
        return [ApiKey.model_validate(item) for item in data.get("keys", [])]

    from typing import Any

    def create_api_key(self, name: str, scopes: list[str]) -> dict[str, Any]:
        """Create a new API key. Returns dict with ApiKey object and the 'key' (secret)."""
        data = self.http.request("POST", "/v1/org/api-keys", json={"name": name, "scopes": scopes})
        api_key = ApiKey.model_validate(data["apiKey"])
        return {"api_key": api_key, "key": data["key"]}

    def revoke_api_key(self, key_id: str) -> None:
        """Revoke an API key."""
        self.http.request("DELETE", f"/v1/org/api-keys/{key_id}")

File: sdk/typescript/src/resources/org.ts
import { HttpClient } from "../http.js";
import { RosetClientConfig, Tenant, Member, Invitation, ApiKey } from "../types.js";

/**
 * Organization/Tenant Resource
 * Manage workspace settings, members, and API keys
 */
export class OrgResource {
  constructor(
    private readonly http: HttpClient,
    private readonly config: RosetClientConfig
  ) {}

  /**
   * Get current tenant details
   */
  async getTenant(): Promise<Tenant> {
    const { tenant } = await this.http.get<{ tenant: Tenant }>("/v1/org");
    return tenant;
  }

  /**
   * List workspace members
   */
  async listMembers(): Promise<Member[]> {
    const { members } = await this.http.get<{ members: Member[] }>("/v1/org/members");
    return members;
  }

  /**
   * Invite a new member
   */
  async inviteMember(params: { email: string; role: string }): Promise<Invitation> {
    const { invitation } = await this.http.post<{ invitation: Invitation }>(
      "/v1/org/invites",
      params
    );
    return invitation;
  }

  /**
   * List API keys
   */
  async listApiKeys(): Promise<ApiKey[]> {
    const { keys } = await this.http.get<{ keys: ApiKey[] }>("/v1/org/api-keys");
    return keys;
  }

  /**
   * Create a new API key
   */
  async createApiKey(params: { name: string; scopes: string[] }): Promise<ApiKey & { key: string }> {
    const { apiKey, key } = await this.http.post<{ apiKey: ApiKey; key: string }>(
      "/v1/org/api-keys",
      params
    );
    // Return the full object including the secret key (only shown once)
    return { ...apiKey, key };
  }

  /**
   * Revoke an API key
   */
  async revokeApiKey(id: string): Promise<void> {
    await this.http.delete(`/v1/org/api-keys/${id}`);
  }
}
